### ISSUE TITLE
Context cancellation can cause subsequent operation to fail

### ISSUE THREAD

**Issue by niklasj** [mattn/go-sqlite3#488]
Cancelling a context after ExecContext has returned (like with the usual `defer cancel()`) may cause an unwanted SQLITE_INTERRUPT, which in turn can cause an error in a subsequent operation.

The problem seems to be a race between `ctx.Done()` and `done` in https://github.com/mattn/go-sqlite3/blob/master/sqlite3.go#L1170
```
        go func(db *C.sqlite3) {
                select {
                case <-ctx.Done():
                        C.sqlite3_interrupt(db)
                case <-done:
                }
        }(s.c.db)
```

**Comment by mattn** [comment]
Could you please explain how to reproduce?

2017/11/21 午後9:31 "Niklas Janlert" <notifications@github.com>:

**Comment by niklasj** [comment]
I've created a pull requests which fixes the issue and also includes a test that shows how the problem can be reproduced.

**Comment by mattn** [comment]
closed by #489
